src/compiler/program.ts

```diff
@@ -205,13 +205,15 @@ namespace ts {
     }
 
     export function getPreEmitDiagnostics(program: Program, sourceFile?: SourceFile, cancellationToken?: CancellationToken): Diagnostic[] {
-        let diagnostics = program.getOptionsDiagnostics(cancellationToken).concat(
-            program.getSyntacticDiagnostics(sourceFile, cancellationToken),
-            program.getGlobalDiagnostics(cancellationToken),
-            program.getSemanticDiagnostics(sourceFile, cancellationToken));
+        let diagnostics = [
+            ...program.getOptionsDiagnostics(cancellationToken),
+            ...program.getSyntacticDiagnostics(sourceFile, cancellationToken),
+            ...program.getGlobalDiagnostics(cancellationToken),
+            ...program.getSemanticDiagnostics(sourceFile, cancellationToken)
+        ];
 
         if (program.getCompilerOptions().declaration) {
-            diagnostics = diagnostics.concat(program.getDeclarationDiagnostics(sourceFile, cancellationToken));
+            diagnostics = [...diagnostics, ...program.getDeclarationDiagnostics(sourceFile, cancellationToken)];

```

Comment:
Unfortunately, we're using three different kinds of concatenation in the codebase: ES6 `...`, ES5 `.concat`, and ES3 `ts.concatenate`. The `.concat()` method won't typecheck against readonly arrays (#17076) but does the same thing as `...`. We should really prefer `ts.concatenate` which won't allocate output if the inputs are empty, but in this case we're currently returning a mutable `Diagnostic[]`, so we can't use any function that might reuse its readonly inputs -- it has to be a freshly-allocated output.